Vectorize CPU ATen mean kernel for BF16 & FP16 dtypes

sanchitintel opened the pull request:
## Summary
Since #97351, CPU ATen kernel for `mean` for BF16 & FP16 dtypes has been unvectorized (it's not even implicitly vectorized).

This PR vectorizes `mean` for BF16 & FP16 on CPU in a `cast_fp32 -> sum -> div -> cast_bf16_or_fp16` fashion.

The perf benefit would be especially pronounced on machines with `AVX512_BF16` and/or `AVX512_FP16` ISA support.

## Benchmarking data for BF16 (collected before & after the change in this PR)

**Machine:** Intel&reg; Xeon&reg; (4th generation series, formerly codenamed Sapphire Rapids) Platinum 8468H
One socket (48 physical cores) - used `numactl --membind=0 --cpunodebind=0`
libtcmalloc & Intel OpenMP were preloaded

Environment variable used -
`KMP_AFFINITY=granularity=fine,compact,1,0 KMP_BLOCKTIME=1 KMP_SETTINGS=1 OMP_NUM_THREADS=48 MKL_NUM_THREADS=48`

**Workload:** E2E performance on BS 32 resnet50 (using BF16 via AMP) inference using oneDNN Graph JIT fuser (`mean` kernel is dispatched to eager mode ATen kernel, and is the bottleneck right now)

| **BEFORE:** Latency with unvectorized mean (lower is better)| **AFTER:** Latency with vectorized mean (lower is better)| Speedup due to vectorizing mean|
|----------------------------|-------------------------|------------|
|                19.1 ms           |                10.8  ms       | latency reduced by ~43.45%      |

**Benchmarking script for BF16 -**

 ```
import time
import torch
import torchvision

# enable oneDNN Graph JIT fuser
torch.jit.enable_onednn_fusion(True)
# AMP for JIT mode is enabled by default, and is divergent with its eager mode counterpart
torch._C._jit_set_autocast_mode(False)

# sample input should be of the same shape as expected inputs
example_input = torch.rand(32, 3, 224, 224)
# Using resnet50 from torchvision in this example for illustrative purposes,
# but the line below can indeed be modified to use custom models as well.
model = getattr(torchvision.models, "resnet50")().eval()

with torch.no_grad(), torch.cpu.amp.autocast(cache_enabled=False, dtype=torch.bfloat16):
    # Conv-BatchNorm folding for CNN-based Vision Models should be done with ``torch.fx.experimental.optimization.fuse`` when AMP is used
    import torch.fx.experimental.optimization as optimization
    # Please note that optimization.fuse need not be called when AMP is not used
    model = optimization.fuse(model)
    model = torch.jit.trace(model, (example_input))
    model = torch.jit.freeze(model)
    # a couple of warm-up runs
    model(example_input)
    model(example_input)
    # speedup would be observed in subsequent runs
    start = time.time()
    model(example_input)
    end = time.time()
    inference_time = (end - start) * 1000
    print("Inference time is ", inference_time)
```



cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10

sanchitintel committed:
Vectorize CPU mean kernel for BF16 dtype

Currently, CPU kernel for mean is unvectorized (it seems it can't even be implicitly vectorized).

sanchitintel committed:
Update comments

sanchitintel committed:
Refactor to use a switch statement

sanchitintel committed:
Revise comments

sanchitintel commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1338,8 +1338,53 @@ TORCH_IMPL_FUNC(mean_out)
+        //  cast_fp32 -> sum -> div -> cast_bf16
+        //
+        // This approach results in one extra pass over all the
+        // elements of the output tensor, but that overhead is amortized by
```
nit:

`output -> `input`

Will change it after all CI checks will complete

sanchitintel committed:
Revised comments

Comments were too verbose & confusing

sanchitintel committed:
Fix typo in comments

sanchitintel committed:
Fix typo in comments

yet again :(

sanchitintel committed:
Fix typo again :(

jgong5 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1338,8 +1336,53 @@ TORCH_IMPL_FUNC(mean_out)
+        // Cast result_mut back to BF16 or FP16.
+        // A digression - the copy argument to at::native::to would be ignored,
+        // so it might as well have been true in both the at::native::to calls here.
+        result_mut = result_mut.to(dtype,
```
ditto

jgong5 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1338,8 +1336,53 @@ TORCH_IMPL_FUNC(mean_out)
+        // some modifications in binary_kernel_reduce_vec(),
+        // TensorIteratorBase::for_each(), and
+        // TensorIteratorBase::serial_for_each().
+        at::sum_out(
+            result_mut, self, opt_dim, keepdim, ScalarType::Float).div_(dim_prod);
```
I guess you can share this line with fp32 path?

sanchitintel replied:
No, because FP32 path is shared by some other dtypes as well. 
BTW, `mean_kernel_impl` is not being used by any dtype now.
Would you recommend removing it from the PyTorch codebase? Thanks!

sanchitintel replied:
Eliminated duplicate code. Thank you!

sanchitintel committed:
Revise code as per review comments

TensorBase::to is being called but I had looked at the definition of at::native::to

sanchitintel commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1338,8 +1336,53 @@ TORCH_IMPL_FUNC(mean_out)
+                                   /*pin_memory=*/false,
+                                   /*non_blocking=*/false,
+                                   /*copy=*/false,
+                                   /*memory_format=*/at::MemoryFormat::Preserve);
```
Oops, I had looked at the definition of `at::native::to` instead of `TensorBase::to`. Fixed it. Thank you!

sanchitintel committed:
Remove redundant comments

sanchitintel committed:
Clarify code-flow-path in comments

sanchitintel committed:
Refactor to eliminate duplicate code

sanchitintel committed:
Move comments

sanchitintel committed:
Update comments on function-call overhead

Eliminating duplicate code led to unnecessary function-call overhead for all dtypes besides FP16 & BF16

sanchitintel committed:
Avoid unnecessary function calls for non-half dtypes

avoid the overhead of unnecessary TensorBase::to calls for non-half dtypes (functionally, they'd have been no-ops).

sanchitintel committed:
Remove redundant code

No code-coverage  (i.e. not being called)

sanchitintel committed:
Revert last change

DEFINE_DISPATCH for mean_stub expects a CPU implementation as well

sanchitintel committed:
Remove mean_kernel_impl

Dispatch mean_stub to a nullptr for CPU

sanchitintel committed:
Change dispatch macro used for mean_stub

No need to use nullptr since I had added a relevant macro in DispatchStub

sanchitintel committed:
Use REGISTER_DISPATCH(mean_stub, nullptr);

REGISTER_NO_CPU_DISPATCH is not working. Should remove it

sanchitintel commented:
@pytorchbot merge

sanchitintel commented:
Hi @ezyang, can you please help review this PR? It requires approval from a core maintainer.
If you'd have availability to review, please ignore the numerous commits & please only look at the changes introduced in this PR.

Thank you! :)

sanchitintel commented:
Hi @malfet, can you please help review this PR?
eager-mode `mean` kernel for BF16/FP16 datatypes has been unvectorized since March, and since it's called by `adaptive_avg_pool2d` in models such as resnet50, their performance with AMP is poor.
This PR fixes the issue by vectorizing `mean` computation for BF16 & FP16 dtypes as well.

Thanks!

malfet commented:
LGTM

malfet commented:
@pytorchbot merge
